File: java/900/LeetCode0983.java
public class LeetCode0983 {
    public int mincostTickets(int[] days, int[] costs) {
        int[][] dp = new int[days.length][3];
        return helper(days, costs, dp, 0, 0);
    }
    public int helper(int[] days,int[] costs,int[][] dp,int index,int cost) {
        if (index >= days.length) {
            return cost;
        }
        if (dp[index][0] == 0) {
            dp[index][0] = helper(days, costs, dp, index + 1, costs[0]);
        }
        if (dp[index][1] == 0) {
            int newi = index;
            for (;newi < days.length && days[newi] < days[index] + 7;++newi);
            dp[index][1] = helper(days, costs, dp, newi, costs[1]);
        }
        if (dp[index][2] == 0){
            int newi = index;
            for (;newi < days.length && days[newi] < days[index] + 30;++newi);
            dp[index][2] = helper(days, costs, dp, newi, costs[2]);
        }
        int daycost = Math.min(dp[index][0], Math.min(dp[index][1], dp[index][2]));
        return daycost + cost;
    }
}

File: cpp/L900/LeetCode0983.cc
#include <vector>
using std::vector;

class Solution {
public:
    int mincostTickets(vector<int>& days, vector<int>& costs) {
        int N = days.size();
        vector<int> totals(N + 1, 0);
        for (int i = 1;i <= N;++i) {
            totals[i] = totals[i - 1] + costs[0];
            for (int j = i;j - 1 >= 0 && days[i - 1] - days[j - 1] + 1 <= 30;--j) {
                int dayCount = days[i - 1] - days[j - 1] + 1;
                if (dayCount <= 7) {
                    totals[i] = std::min(totals[i], totals[j - 1] + costs[1]);
                } 
                totals[i] = std::min(totals[i], totals[j - 1] + costs[2]);
            }
        }
        return totals[N];
    }
};